src/garage/envs/environment.py

```diff
@@ -0,0 +1,189 @@
+"""Base Garage Environment API."""
+
+import abc
+
+
+class Environment(abc.ABC):
+    """The main API for garage environments.
+
+    The main public API methods are:
+
+        step
+        reset
+        render
+        visualize
+        close
+
+    And set the following attributes:
+
+        action_space: The Space object corresponding to valid actions
+        observation_space: The Space object corresponding to valid observations
+        reward_range: A tuple corresponding to the min and max possible rewards
+
+    Make sure your environment is pickle-able:
+        Garage pickles the environment via the `pickle` or `cloudpickle` module
+        to save snapshots of the experiment. However, some environments may
+        contain attributes that are not pickle-able (e.g. a client-server
+        connection). In such cases, override `__getstate__()` and
+        `__setstate()__` to add your custom pickle logic.
+
+    Note: a default reward range set to [-inf,+inf] already exists. Set it if
+        you want a narrower range.
+
+    """
+
+    # Set this in SOME subclasses
+    metadata = {'render.modes': []}
+    reward_range = (-float('inf'), float('inf'))
+    spec = None
```

Comment:
what is this?

Reply:
For garage, only `spec` from `gym.Env` is used. In fact `GarageEnv` provides its own way of returning `spec`. I will only keep `action_space`, `observation_space`, `spec` and `render_modes` as the abstract attributes for this API.